{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### In this notebook are the diferent scritps I use to get the graphics we need to understand the behavior of the genomes and chromosomas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import altair as alt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "class GenomaScatter700():\n",
    "\n",
    "    def decodeTranscript(self,x):\n",
    "        '''\n",
    "        This function decode the transcript putting the right name of the gene\n",
    "        '''\n",
    "        gen = x.split('_')[0]\n",
    "        gen_d = {'chr17':'_BRCA1', 'chr16':'_PALB2','chr13':'_BRCA2','chr11':'_ATM','chr2':'_MSH2', }\n",
    "        return gen + gen_d[gen]\n",
    "\n",
    "\n",
    "    def clean(self):\n",
    "        df = pd.read_csv('scater_dataOK_revisado.txt', delimiter='\\t')                     # A: Convert the file into a DF\n",
    "        df.rename(columns={'add2':'Tries','add3':'Genes_N'},inplace=True)                  # A: Change the name of the columns\n",
    "        df['Genes'] = df['transcript'].apply(lambda x: self.decodeTranscript(x))           # A: Creo columna con valores de transcript\n",
    "\n",
    "        # Convert the data into categorical data reduce the use of memory from 67.7 kb kb to 41.3kb\n",
    "        df['sampleID'] = df['sampleID'].astype('category')\n",
    "        df['add1'] = df['add1'].astype('category')\n",
    "        df['transcript'] = df['transcript'].astype('category')\n",
    "        df['Genes'] = df['Genes'].astype('category')\n",
    "\n",
    "        return df\n",
    "\n",
    "\n",
    "    def scatter_size(self, df):\n",
    "        # Conditional variables\n",
    "        click = alt.selection_multi(encodings=['color'])                                   # U: Can select the genoma and filter the graph\n",
    "        \n",
    "        # Chart variables\n",
    "        x_circle = alt.X('sampleID:O', axis = alt.Axis(title = 'sampleID', grid=True))   # U: The X with de Label Angel\n",
    "        y_circle = alt.Y('TPM', axis = alt.Axis(title = 'TPM', grid=True))\n",
    "        \n",
    "        color_circle = alt.condition(click, alt.Color(shorthand='Genes:N'), alt.value('lightgray'))                         # U: The color variable without the legend     \n",
    "        size_circle = alt.Size('transcript:N', scale=alt.Scale(range=[150, 200]),legend=None)           # U: The size with the range of scale\n",
    "        \n",
    "\n",
    "        \n",
    "        # Chart base\n",
    "        circle = alt.Chart(df).mark_circle().encode(x = x_circle, y = y_circle, color = color_circle, size = size_circle     # U: The basic chart\n",
    "                                          ).properties(height=720                                               # U: The heigth and width\n",
    "                                          ).add_selection(click)\n",
    "        line = alt.Chart(df).mark_line().encode(x = x_circle, y = y_circle, color = color_circle, #size = size_circle     # U: The basic chart\n",
    "                                          ).properties(height=720                                               # U: The heigth and width\n",
    "                                          )\n",
    "\n",
    "        \n",
    "        file = circle + line\n",
    "        \n",
    "        \n",
    "        return file.save('./Plotts/Scatter_Dinamico_700.html') \n",
    "    \n",
    "        \n",
    "    def graf(self):\n",
    "        df = self.clean()\n",
    "        return self.scatter_size(df)\n",
    "\n",
    "g = GenomaScatter700()\n",
    "g.graf()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "class GenomaScatter600():\n",
    "\n",
    "    def decodeTranscript(self,x):\n",
    "        '''\n",
    "        This function decode the transcript putting the right name of the gene\n",
    "        '''\n",
    "        gen = x.split('_')[0]\n",
    "        gen_d = {'chr17':'_BRCA1', 'chr16':'_PALB2','chr13':'_BRCA2','chr11':'_ATM','chr2':'_MSH2', }\n",
    "        return gen + gen_d[gen]\n",
    "\n",
    "\n",
    "    def clean(self):\n",
    "        df = pd.read_csv('scater_dataOK_revisado.txt', delimiter='\\t')                     # A: Convert the file into a DF\n",
    "        df.rename(columns={'add2':'Tries','add3':'Genes_N'},inplace=True)                  # A: Change the name of the columns\n",
    "        df['Genes'] = df['transcript'].apply(lambda x: self.decodeTranscript(x))           # A: Creo columna con valores de transcript\n",
    "        df['Tries'] = df['Tries'].apply(lambda x: str('{:02d}'.format((x))))\n",
    "        df['Tries'] = df['sampleID'].str.cat(df['Tries'],sep=\" \")           # A:\n",
    "\n",
    "        # Convert the data into categorical data reduce the use of memory from 67.7 kb kb to 41.3kb\n",
    "        df['sampleID'] = df['sampleID'].astype('category')\n",
    "        df['add1'] = df['add1'].astype('category')\n",
    "        df['transcript'] = df['transcript'].astype('category')\n",
    "        df['Genes'] = df['Genes'].astype('category')\n",
    "\n",
    "        return df\n",
    "\n",
    "\n",
    "    def scatter_size(self, df):\n",
    "        # Conditional variables\n",
    "        click = alt.selection_multi(encodings=['color'])                                   # U: Can select the genoma and filter the graph\n",
    "        \n",
    "        # Chart variables\n",
    "        x_circle = alt.X('Tries:O', axis = alt.Axis(title = 'Tries', grid=True))   # U: The X with de Label Angel\n",
    "        y_circle = alt.Y('TPM', axis = alt.Axis(title = 'TPM', grid=True))\n",
    "        \n",
    "        color_circle = alt.condition(click, alt.Color(shorthand='Genes:N'), alt.value('lightgray'))                         # U: The color variable without the legend     \n",
    "        size_circle = alt.Size('transcript:N', scale=alt.Scale(range=[150, 200]),legend=None)           # U: The size with the range of scale\n",
    "        \n",
    "\n",
    "        \n",
    "        # Chart base\n",
    "        circle = alt.Chart(df).mark_circle().encode(x = x_circle, y = y_circle, color = color_circle, size = size_circle     # U: The basic chart\n",
    "                                          ).properties(height=720                                               # U: The heigth and width\n",
    "                                          ).add_selection(click)\n",
    "        line = alt.Chart(df).mark_line().encode(x = x_circle, y = y_circle, color = color_circle, #size = size_circle     # U: The basic chart\n",
    "                                          ).properties(height=720                                               # U: The heigth and width\n",
    "                                          )\n",
    "\n",
    "        \n",
    "        file = circle + line\n",
    "        \n",
    "        \n",
    "        return file.save('./Plotts/Scatter_Dinamico_600.html') \n",
    "    \n",
    "        \n",
    "    def graf(self):\n",
    "        df = self.clean()\n",
    "        return self.scatter_size(df)\n",
    "\n",
    "g = GenomaScatter600()\n",
    "g.graf()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "class GenomaScatter500():\n",
    "\n",
    "    def decodeTranscript(self,x):\n",
    "        '''\n",
    "        This function decode the transcript putting the right name of the gene\n",
    "        '''\n",
    "        gen = x.split('_')[0]\n",
    "        gen_d = {'chr17':'_BRCA1', 'chr16':'_PALB2','chr13':'_BRCA2','chr11':'_ATM','chr2':'_MSH2', }\n",
    "        return gen + gen_d[gen]\n",
    "\n",
    "\n",
    "    def clean(self):\n",
    "        df = pd.read_csv('scater_dataOK_revisado.txt', delimiter='\\t')                     # A: Convert the file into a DF\n",
    "        df.rename(columns={'add2':'Tries','add3':'Genes_N'},inplace=True)                  # A: Change the name of the columns\n",
    "        df['Genes'] = df['transcript'].apply(lambda x: self.decodeTranscript(x))           # A: Creo columna con valores de transcript\n",
    "\n",
    "        # Convert the data into categorical data reduce the use of memory from 67.7 kb kb to 41.3kb\n",
    "        df['sampleID'] = df['sampleID'].astype('category')\n",
    "        df['add1'] = df['add1'].astype('category')\n",
    "        df['transcript'] = df['transcript'].astype('category')\n",
    "        df['Genes'] = df['Genes'].astype('category')\n",
    "\n",
    "        return df\n",
    "\n",
    "\n",
    "    def scatter_size(self, df):\n",
    "        # Conditional variables\n",
    "        click = alt.selection_multi(encodings=['color'])                                   # U: Can select the genoma and filter the graph\n",
    "        \n",
    "        # Chart variables\n",
    "        x_circle = alt.X('TPM', axis = alt.Axis(title = 'TPM', grid=True))   # U: The X with de Label Angel\n",
    "        y_circle = alt.Y('sampleID', axis = alt.Axis(title = 'Patients', grid=True))\n",
    "        \n",
    "        color_circle = alt.Color(shorthand='Genes:N', legend=None)                         # U: The color variable without the legend     \n",
    "        size_circle = alt.Size('transcript:N', scale=alt.Scale(range=[150, 200]))           # U: The size with the range of scale\n",
    "        \n",
    "        \n",
    "        # Legend variables\n",
    "        y_legend_c = alt.Y('Genes:N', axis=alt.Axis(title='Genes'))                            # U: Y value for the legend\n",
    "        color_legend_c = alt.condition(click, 'Genes:N', alt.value('lightgray'), legend=None)  # U: The color with the condition.\n",
    "        \n",
    "        \n",
    "        # Chart base\n",
    "        chart = alt.Chart(df).mark_circle().encode(x = x_circle, y = y_circle, color = color_circle, size = size_circle     # U: The basic chart\n",
    "                                          ).properties(width=720, height=1280                                               # U: The heigth and width\n",
    "                                          ).transform_filter(click)                                                         # U: A transform filter that let me alterate the graf by the selection in the legend.\n",
    "\n",
    "        \n",
    "        # Legend base \n",
    "        legend = alt.Chart(df).mark_circle().encode(y = y_legend_c, color= color_legend_c, size=alt.value(250)\n",
    "                                             ).add_selection(click)                                                         # U: The click to select the genomas\n",
    "        \n",
    "        file = chart | legend\n",
    "        \n",
    "        \n",
    "        return file.save('./Plotts/Scatter_Dinamico_500.html') \n",
    "    \n",
    "        \n",
    "    def graf(self):\n",
    "        df = self.clean()\n",
    "        return self.scatter_size(df)\n",
    "\n",
    "g = GenomaScatter500()\n",
    "g.graf()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [],
   "source": [
    "class GenomaScatter302():\n",
    "\n",
    "    def decodeTranscript(self,x):\n",
    "        '''\n",
    "        This function decode the transcript putting the right name of the gene\n",
    "        '''\n",
    "        gen = x.split('_')[0]\n",
    "        gen_d = {'chr17':'_BRCA1', 'chr16':'_PALB2','chr13':'_BRCA2','chr11':'_ATM','chr2':'_MSH2', }\n",
    "        return gen + gen_d[gen]\n",
    "\n",
    "\n",
    "    def clean(self):\n",
    "        df = pd.read_csv('scater_dataOK_revisado.txt', delimiter='\\t')                     # A: Convert the file into a DF\n",
    "        df.rename(columns={'add2':'Tries','add3':'Genes_N'},inplace=True)                  # A: Change the name of the columns\n",
    "        df['Genes'] = df['transcript'].apply(lambda x: self.decodeTranscript(x))           # A: Creo columna con valores de transcript\n",
    "\n",
    "        # Convert the data into categorical data reduce the use of memory from 67.7 kb kb to 41.3kb\n",
    "        df['sampleID'] = df['sampleID'].astype('category')\n",
    "        df['add1'] = df['add1'].astype('category')\n",
    "        df['transcript'] = df['transcript'].astype('category')\n",
    "        df['Genes'] = df['Genes'].astype('category')\n",
    "\n",
    "        return df\n",
    "\n",
    "\n",
    "    def scatter_size(self, df):\n",
    "        \n",
    "        l_chart = []\n",
    "        \n",
    "        \n",
    "        click = alt.selection_multi(encodings=['color'])\n",
    "          \n",
    "        for x in range(1,5):\n",
    "            # Bases for loop\n",
    "            x = str('{:02d}'.format(x))\n",
    "            b_df = df[df['sampleID']=='S{}'.format(x)]\n",
    "            \n",
    "            \n",
    "            # Chart conditional\n",
    "            \n",
    "            \n",
    "            \n",
    "            # Chart variables\n",
    "            x_c = alt.X('Tries', axis = alt.Axis(title = 'Tries of S{}'.format(x)), scale= alt.Scale(domain=[0, 14]) )\n",
    "            y_c = alt.Y('TPM', axis = alt.Axis(title = 'TPM'), scale= alt.Scale(domain=[0, 350000]) )\n",
    "            \n",
    "            color_c = alt.condition(click, alt.Color(shorthand='Genes:N'), alt.value('lightgray'))\n",
    "            size_c = alt.Size('transcript:N', scale=alt.Scale(range=[50, 150]))\n",
    "            \n",
    "            \n",
    "            # Chart base\n",
    "            b_ch = alt.Chart(b_df).mark_circle().encode(x= x_c, y=y_c, color=color_c, size=size_c\n",
    "                                               ).properties(width=270, height=180\n",
    "                                               ).add_selection(click)\n",
    "            \n",
    "            \n",
    "            \n",
    "            l_chart.append(b_ch)\n",
    "            \n",
    "        f_chart = l_chart[0] | l_chart[1] | l_chart[2] | l_chart[3] \n",
    "        \n",
    "        f2_chart = f_chart & f_chart & f_chart & f_chart \n",
    "        \n",
    "        f2_chart.save('Scatter_Dinamico_302.html')\n",
    "        \n",
    "        return\n",
    "    \n",
    "        \n",
    "    def graf(self):\n",
    "        df = self.clean()\n",
    "        return self.scatter_size(df)\n",
    "\n",
    "g = GenomaScatter302()\n",
    "g.graf()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "class GenomaScatter500():\n",
    "\n",
    "    def decodeTranscript(self,x):\n",
    "        '''\n",
    "        This function decode the transcript putting the right name of the gene\n",
    "        '''\n",
    "        gen = x.split('_')[0]\n",
    "        gen_d = {'chr17':'_BRCA1', 'chr16':'_PALB2','chr13':'_BRCA2','chr11':'_ATM','chr2':'_MSH2', }\n",
    "        return gen + gen_d[gen]\n",
    "\n",
    "\n",
    "    def clean(self):\n",
    "        df = pd.read_csv('scater_dataOK_revisado.txt', delimiter='\\t')                     # A: Convert the file into a DF\n",
    "        df.rename(columns={'add2':'Tries','add3':'Genes_N'},inplace=True)                  # A: Change the name of the columns\n",
    "        df['Genes'] = df['transcript'].apply(lambda x: self.decodeTranscript(x))           # A: Creo columna con valores de transcript\n",
    "\n",
    "        # Convert the data into categorical data reduce the use of memory from 67.7 kb kb to 41.3kb\n",
    "        df['sampleID'] = df['sampleID'].astype('category')\n",
    "        df['add1'] = df['add1'].astype('category')\n",
    "        df['transcript'] = df['transcript'].astype('category')\n",
    "        df['Genes'] = df['Genes'].astype('category')\n",
    "\n",
    "        return df\n",
    "\n",
    "\n",
    "    def scatter_size(self, df):\n",
    "        # Conditional variables\n",
    "        click = alt.selection_multi(encodings=['color'])                                   # U: Can select the genoma and filter the graph\n",
    "        \n",
    "        # Chart variables [Sample con X]\n",
    "        x_circle = alt.X('sampleID', axis = alt.Axis(title = 'Patients', labelAngle=-90))  # U: The X with de Label Angel\n",
    "        y_circle = alt.Y('TPM', axis = alt.Axis(title = 'TPM')) \n",
    "        \n",
    "        # Chart variables [Sample con Y]\n",
    "        #x_circle = alt.X('TPM', axis = alt.Axis(title = 'TPM', grid=True))   # U: The X with de Label Angel\n",
    "        #y_circle = alt.Y('sampleID', axis = alt.Axis(title = 'Patients', grid=True))\n",
    "        \n",
    "        \n",
    "        color_circle = alt.Color(shorthand='Genes:N', legend=None)                         # U: The color variable without the legend     \n",
    "        size_circle = alt.Size('transcript:N', scale=alt.Scale(range=[150, 200]))           # U: The size with the range of scale\n",
    "        \n",
    "        \n",
    "        # Legend variables\n",
    "        y_legend_c = alt.Y('Genes:N', axis=alt.Axis(title='Genes'))                            # U: Y value for the legend\n",
    "        color_legend_c = alt.condition(click, 'Genes:N', alt.value('lightgray'), legend=None)  # U: The color with the condition.\n",
    "        \n",
    "        \n",
    "        # Chart base\n",
    "        chart = alt.Chart(df).mark_circle().encode(x = x_circle, y = y_circle, color = color_circle, size = size_circle     # U: The basic chart\n",
    "                                          ).properties(width=1280, height=720                                               # U: The heigth and width\n",
    "                                          ).transform_filter(click)                                                         # U: A transform filter that let me alterate the graf by the selection in the legend.\n",
    "\n",
    "        \n",
    "        # Legend base \n",
    "        legend = alt.Chart(df).mark_circle().encode(y = y_legend_c, color= color_legend_c, size=alt.value(250)\n",
    "                                             ).add_selection(click)                                                         # U: The click to select the genomas\n",
    "        \n",
    "        file = chart | legend\n",
    "        \n",
    "        file.save('./Plotts/Scatter_Dinamico_401.html') \n",
    "        return #file\n",
    "    \n",
    "        \n",
    "        \n",
    "    def heatmap_size(self, df):\n",
    "                   \n",
    "            \n",
    "        heatmap = alt.Chart(df).mark_rect().encode(\n",
    "            alt.X('sampleID:O'),\n",
    "            alt.Y('Tries:O'),\n",
    "            alt.Color('Genes_N:Q')\n",
    "        ).properties(width=1280, height=1080)\n",
    "        \n",
    "        \n",
    "        return heatmap\n",
    "        \n",
    "    def graf(self):\n",
    "        df = self.clean()\n",
    "        scatter = self.scatter_size(df)\n",
    "        heatmap = self.heatmap_size(df)\n",
    "        \n",
    "        return heatmap\n",
    "\n",
    "g = GenomaScatter500()\n",
    "#g.graf()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "class GenomaScatter299():\n",
    "\n",
    "    def decodeTranscript(self,x):\n",
    "        '''\n",
    "        This function decode the transcript putting the right name of the gene\n",
    "        '''\n",
    "        gen = x.split('_')[0]\n",
    "        gen_d = {'chr17':'_BRCA1', 'chr16':'_PALB2','chr13':'_BRCA2','chr11':'_ATM','chr2':'_MSH2', }\n",
    "        return gen + gen_d[gen]\n",
    "\n",
    "\n",
    "    def clean(self):\n",
    "        df = pd.read_csv('scater_dataOK_revisado.txt', delimiter='\\t')                     # A: Convert the file into a DF\n",
    "        df.rename(columns={'add2':'Tries','add3':'Genes_N'},inplace=True)                  # A: Change the name of the columns\n",
    "        df['Genes'] = df['transcript'].apply(lambda x: self.decodeTranscript(x))           # A: Creo columna con valores de transcript\n",
    "\n",
    "        # Convert the data into categorical data reduce the use of memory from 67.7 kb kb to 41.3kb\n",
    "        df['sampleID'] = df['sampleID'].astype('category')\n",
    "        df['add1'] = df['add1'].astype('category')\n",
    "        df['transcript'] = df['transcript'].astype('category')\n",
    "        df['Genes'] = df['Genes'].astype('category')\n",
    "\n",
    "        return df\n",
    "\n",
    "\n",
    "    def scatter_size(self, df):\n",
    "        # Conditional variables\n",
    "        click = alt.selection_multi(encodings=['color'])                                   # U: Can select the genoma and filter the graph\n",
    "        \n",
    "        # Chart variables\n",
    "        x_circle = alt.X('sampleID', axis = alt.Axis(title = 'Patients', labelAngle=-90))  # U: The X with de Label Angel\n",
    "        y_circle = alt.Y('TPM', axis = alt.Axis(title = 'TPM')) \n",
    "        color_circle = alt.condition(click, alt.Color(shorthand='Genes:N'), alt.value('lightgray'))  # The color with teh click condition     \n",
    "        size_circle = alt.Size('transcript:N', scale=alt.Scale(range=[50, 200]))                     # The size with the range of scale\n",
    "\n",
    "        chart = alt.Chart(df).mark_circle().encode(x = x_circle, y = y_circle, color = color_circle, size = size_circle     # The basic chart\n",
    "                                          ).properties(width=1280, height=720                                               # The heigth and width\n",
    "                                          ).add_selection(click)                                                            # The selection\n",
    "\n",
    "        chart.save('Scatter_Dinamico.html') \n",
    "        return \n",
    "    \n",
    "        \n",
    "    def graf(self):\n",
    "        df = self.clean()\n",
    "        return self.scatter_size(df)\n",
    "\n",
    "g = GenomaScatter299()\n",
    "#g.graf()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def graf(df):\n",
    "    sns.set(style='darkgrid')                                # A: Seteo el color de fondo\n",
    "    fig = plt.figure(figsize=(1440/96, 600/96), dpi=96)      # A: Doy tamaño a la figura\n",
    "\n",
    "    sns.scatterplot(x='sampleID',                            # A: Creo el plot\n",
    "                    y= 'value', \n",
    "                    hue= 'Muestras',\n",
    "                    data=df)\n",
    "    \n",
    "    plt.xlabel('Muestras')                                   # A: Cambio el nombre de la etiqueta\n",
    "    plt.xticks(rotation=45)                                  # A: Roto las etiquetas para que sean legibles\n",
    "    plt.xlim(-1,96)                                          # A: Limito el eje x para que tome todo el gráfico\n",
    "    \n",
    "    plt.ylabel('Valores')                                    # A: Cambio el nombre de la etiqueta\n",
    "    \n",
    "    plt.legend(bbox_to_anchor=(0., 1.02, 1., .102), loc='lower left', ncol=5,    # A: Coloco la leyenda arriba\n",
    "           mode=\"expand\", borderaxespad=0., ) \n",
    "     \n",
    "    plt.savefig('ejemplo_1', dpi=96)                         # A: Salvo el plot\n",
    "    \n",
    "#graf(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def scatter_line(df):\n",
    "\n",
    "    # Conditional Variables\n",
    "    brush = alt.selection_interval(encodings=['x'])  # U: Active on the top panel \n",
    "    click = alt.selection_multi(encodings=['color']) # U: Active on the bottom panel \n",
    "    \n",
    "    # Chart Variables\n",
    "    x_scatter = alt.X('sampleID', axis = alt.Axis(title = 'Muestras', grid = True, labelAngle=-45))      # U: X axis with the data, title, grid and the agle of label\n",
    "    y_scatter = alt.Y('value', axis = alt.Axis(title = 'Valores', grid = True))                          # U: Y axis with the data, title and grid.\n",
    "    color_scatter = alt.condition(brush, alt.Color(shorthand='Muestras:N'), alt.value('lightgray'))      # U: Color, with condition, the data, and the value of the not selected data\n",
    "    \n",
    "    x_bar = alt.X('value', axis = alt.Axis(title = 'Valores', grid = True))\n",
    "    y_bar = alt.Y('transcript', axis = alt.Axis(title = '¿Genomas?', grid = False))   \n",
    "    color_bar = alt.condition(click, alt.Color(shorthand='transcript:N'), alt.value('lightgray'))\n",
    "    \n",
    "    # Charts\n",
    "    scatter = alt.Chart(df).mark_point().encode(x = x_scatter, y = y_scatter,  color = color_scatter     # U: The basic data of the chart\n",
    "                                       ).properties(width=1280                                           # U: The width, and can go other propertis\n",
    "                                       ).add_selection(brush                                             # U: The selection for this chart\n",
    "                                       ).transform_filter(click)                                         # U: The way this chart can be transformed\n",
    "\n",
    "    bar = alt.Chart(df).mark_bar().encode(x = x_bar, y = y_bar,  color = color_bar \n",
    "                                     ).properties(width=600\n",
    "                                     ).add_selection(click\n",
    "                                     ).transform_filter(brush)\n",
    "    \n",
    "    chart = alt.vconcat(scatter,bar)                                                                     # U: The position of the charts, from up to bottom\n",
    "    chart.save('Scatter_Bar_Dinamico.html')                                                              # A: Saving the chart\n",
    "    \n",
    "    return chart\n",
    "\n",
    "#scatter_line(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Funcion para graficar los Genomas\n",
    "##### Permite:\n",
    "- Hacer un gráfico plano con Seaborn donde se muestran los distintos experimentos y sus resultados. \n",
    "- Hacer un gráfico interactivo donde se muestran los experimentos, sus resultados y se puede filtrar por genoma. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Genoma():\n",
    "    \n",
    "    # ================= Función de seleccion de gráfico =================\n",
    "    def type(self):\n",
    "        \n",
    "        for i in range(3):\n",
    "            \n",
    "            g_type = float(input('Seleccione el tipo de gráfico que desea realizar: \\n[1] Grafico plano con Seaborn \\n[2] Grafico interactivo con Altair\\n[3] Ambos\\n'))\n",
    "\n",
    "            if g_type == 1:\n",
    "                print('Eligió \"Grafico plano con Seaborn\"\\nConsulte la carpeta que el gráfico debe encontrarse allí')\n",
    "                return g_type\n",
    "            elif g_type == 2:\n",
    "                print('Eligió \"Grafico interactivo con Altair\"\\nConsulte la carpeta que el gráfico debe encontrarse allí')\n",
    "                return g_type\n",
    "            elif g_type == 3:\n",
    "                print('Eligió \"Ambos\"\\nConsulte la carpeta que los gráfico deben encontrarse allí')\n",
    "                return g_type\n",
    "            else:\n",
    "                print('Elija una opción válida\\n')\n",
    "                \n",
    "        print('Lo siento, ninguna de las opciones fue válida')\n",
    "    \n",
    "    \n",
    "    # ================ Función de limpieza de Data Frame ================\n",
    "    def clean(self):\n",
    "        df = pd.read_csv('scater_dataOK_revisado.txt', delimiter='\\t')            # A: Tomo el archivo \n",
    "        df.columns = [\"sampleID\", \"add1\", \"transcript\", \"add2\", \"TPM\", \"add3\"]    # A: Doy nombre a las columnas\n",
    "        df['addT'] = df['transcript'].apply(lambda x: x[-1])                      # A: Creo columna con valores de transcript\n",
    "        df['TPM'] = df['TPM'].div(10000)                                          # A: Reduzco el valor para que entre en el gráfico  \n",
    "\n",
    "        df = df.melt(id_vars=['sampleID','transcript'], var_name='Muestras',      # A: Armo el DF para poder graficar\n",
    "                     value_vars=['add2','add3','addT','TPM'])\n",
    "\n",
    "\n",
    "        return df\n",
    "\n",
    "    \n",
    "    # ==================== Función de Grafico Plano =====================\n",
    "    def scatter_seaborn(self, df):\n",
    "        sns.set(style='darkgrid')                                # A: Seteo el color de fondo\n",
    "        fig = plt.figure(figsize=(1440/96, 600/96), dpi=96)      # A: Doy tamaño a la figura\n",
    "\n",
    "        sns.scatterplot(x='sampleID',                            # A: Creo el plot\n",
    "                        y= 'value', \n",
    "                        hue= 'Muestras',\n",
    "                        data=df)\n",
    "\n",
    "        plt.xlabel('Muestras')                                   # A: Cambio el nombre de la etiqueta\n",
    "        plt.xticks(rotation=45)                                  # A: Roto las etiquetas para que sean legibles\n",
    "        plt.xlim(-1,96)                                          # A: Limito el eje x para que tome todo el gráfico\n",
    "\n",
    "        plt.ylabel('Valores')                                    # A: Cambio el nombre de la etiqueta\n",
    "\n",
    "        plt.legend(bbox_to_anchor=(0., 1.02, 1., .102), loc='lower left', ncol=5,    # A: Coloco la leyenda arriba\n",
    "               mode=\"expand\", borderaxespad=0., ) \n",
    "\n",
    "        plt.savefig('Scatter_Bar_Plano', dpi=96)                 # A: Salvo el plot\n",
    "\n",
    "        return\n",
    "\n",
    "    \n",
    "    # ================= Función de Grafico Interactivo ==================\n",
    "    def scatter_altair(self, df):\n",
    "\n",
    "        # Conditional Variables\n",
    "        brush = alt.selection_interval(encodings=['x'])  # U: Active on the top panel \n",
    "        click = alt.selection_multi(encodings=['color']) # U: Active on the bottom panel \n",
    "\n",
    "        # Chart Variables\n",
    "        x_scatter = alt.X('sampleID', axis = alt.Axis(title = 'Muestras', grid = True, labelAngle=-45))      # U: X axis with the data, title, grid and the agle of label\n",
    "        y_scatter = alt.Y('value', axis = alt.Axis(title = 'Valores', grid = True))                          # U: Y axis with the data, title and grid.\n",
    "        color_scatter = alt.condition(brush, alt.Color(shorthand='Muestras:N'), alt.value('lightgray'))      # U: Color, with condition, the data, and the value of the not selected data\n",
    "\n",
    "        x_bar = alt.X('value', axis = alt.Axis(title = 'Valores', grid = True))\n",
    "        y_bar = alt.Y('transcript', axis = alt.Axis(title = 'Genomas', grid = False))   \n",
    "        color_bar = alt.condition(click, alt.Color(shorthand='transcript:N'), alt.value('lightgray'))\n",
    "\n",
    "        # Charts\n",
    "        scatter = alt.Chart(df).mark_point().encode(x = x_scatter, y = y_scatter,  color = color_scatter     # U: The basic data of the chart\n",
    "                                           ).properties(width=1280                                           # U: The width, and can go other propertis\n",
    "                                           ).add_selection(brush                                             # U: The selection for this chart\n",
    "                                           ).transform_filter(click)                                         # U: The way this chart can be transformed\n",
    "\n",
    "        bar = alt.Chart(df).mark_bar().encode(x = x_bar, y = y_bar,  color = color_bar \n",
    "                                         ).properties(width=600\n",
    "                                         ).add_selection(click\n",
    "                                         ).transform_filter(brush)\n",
    "\n",
    "        chart = alt.vconcat(scatter,bar)                                                                     # U: The position of the charts, from up to bottom\n",
    "        chart.save('Scatter_Bar_Dinamico.html')                                                              # A: Saving the chart\n",
    "\n",
    "        return chart\n",
    "\n",
    "    \n",
    "    # ================ Función de Selección de Gráfico ==================\n",
    "    def  graf(self):\n",
    "        \n",
    "        df = self.clean()\n",
    "        sel = self.type()\n",
    "        \n",
    "        if sel == 1:\n",
    "            return self.scatter_seaborn(df)\n",
    "        elif sel == 2:\n",
    "            return self.scatter_altair(df)\n",
    "        elif sel == 3:\n",
    "            return self.scatter_seaborn(df), self.scatter_altair(df)\n",
    "        else:\n",
    "            print('No se como llegaste hasta acá, estás haciendo trampa... ¬¬')\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "g = Genoma()\n",
    "g.graf()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
